kart/dataset3_paths.py

```diff
@@ -0,0 +1,285 @@
+import random
+import subprocess
+from .exceptions import InvalidOperation
+from .serialise_util import (
+    msg_pack,
+    b64encode_str,
+    b64hash,
+    hexhash,
+)
+
+
+# https://datatracker.ietf.org/doc/html/rfc3548.html#section-4
+_BASE64_URLSAFE_ALPHABET = (
+    b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+)
+_BASE64_URLSAFE_ALPHABET_DECODE_MAP = {
+    _BASE64_URLSAFE_ALPHABET[i]: i for i in range(64)
+}
+MAX_B64_INT = 2 ** 29
+
+
+def b64encode_int(integer):
+    """
+    Encodes an integer to a string using exactly five bytes from the urlsafe base64 alphabet.
+    Raises ValueError if the integer is outside the valid range.
+    """
+    if not (-MAX_B64_INT < integer <= MAX_B64_INT):
+        raise ValueError(
+            f"{integer} should be between {-MAX_B64_INT +1} and {MAX_B64_INT}"
+        )
+
+    result = bytearray(bytes(5))
+    for index in range(4, -1, -1):
+        integer, mod = divmod(integer, 64)
+        result[index] = _BASE64_URLSAFE_ALPHABET[mod]
+    return result.decode("ascii")
+
+
+def b64decode_int(s):
+    """
+    Takes a 5-character string, decodes it using base64 as an integer.
+    Reverse of b64encode_int
+    """
+    assert len(s) == 5
+
+    result = 0
+    for i, byt in enumerate(s.encode()):
+        val = _BASE64_URLSAFE_ALPHABET_DECODE_MAP[byt]
+        result += val * 64 ** (4 - i)
+    if result > 2 ** 29:
+        result -= 2 ** 30
+    return result
+
+
+class PathEncoder:
+    """
+    A system for encoding a feature's primary key to a particular path.
+    Originally hex-hash[0:2]/hex-hash[2:4]/base64-encoded-message-packed-primary-key -
+    - that is, 2 levels, and a branch factor of 256 at each level.
+    But, this system proved costly for massive repos and the resulting massive trees.
+    Now we prefer 4 levels, a branch factor of 64, and clustering similar PKs where possible.
+    """
+
+    PATH_STRUCTURE_ITEM = "path-structure.json"
+    PATH_STRUCTURE_PATH = "meta/" + PATH_STRUCTURE_ITEM
+
+    @staticmethod
+    def get(*, scheme, **kwargs):
```

Comment:
Not done - `PathEncoder.get(scheme="int", branches=64, levels=4, encoding="base64")` is readable and works fine - it's easy to find where this get is defined, and no other get method that it would easily be confused with. `for_scheme` makes less sense since it's an implementation detail that the type returned depends on the scheme, as the code changes we might merge those two types into one or add more types that depend on the encoding, or whatever.